packages/core-data/src/selectors.js

```diff
@@ -101,23 +101,43 @@ export function getEntity( state, kind, name ) {
 }
 
 /**
- * Returns the Entity's record object by key.
+ * Returns the Entity's record object by key. Returns `null` if the value is not
+ * yet received, undefined if the value entity is known to not exist, or the
+ * entity object if it exists and is received.
  *
- * @param {Object} state  State tree
- * @param {string} kind   Entity kind.
- * @param {string} name   Entity name.
- * @param {number} key    Record's key
+ * @param {Object}  state State tree
+ * @param {string}  kind  Entity kind.
+ * @param {string}  name  Entity name.
+ * @param {number}  key   Record's key
+ * @param {?Object} query Optional query.
  *
- * @return {Object?} Record.
+ * @return {Object?|undefined} Record.
  */
-export function getEntityRecord( state, kind, name, key ) {
-	return get( state.entities.data, [
+export function getEntityRecord( state, kind, name, key, query ) {
+	// Queried data state is prepopulated for all known entities. If this is not
+	// assigned for the given parameters, then it is known to not exist. Thus, a
+	// return value of `undefined` is used instead of `null` (where `null` is
+	// otherwise used to represent an unknown state).
```

Comment:
Indeed it supports `undefined` is better for data that is not fetched yet and `null` for things we know don't exist.  Why can't we switch it?

Reply:
> Why can't we switch it?

I'm not sure I understood. :) Are you suggesting we change to what I said (undef -> unknown, null -> known nonexistence)? That would be good with me.

Reply:
Yes, that's my suggestion. At least it seems more logical to me but I don't know why @aduth went the other direction initially.

Reply:
Ok, I spent some time looking into this and I think we were misunderstanding the intent.

Returning  `undefined` here doesn't mean the request record  is known to not exist but it instead means the Entity doesn't entity (for example if I request a CPT "book" with id 1, it will return  undefined if the entity doesn't exist  but if the entity  exists and the book is not found or not fetched yet, it will still return null)

I also found that we already return `null` in some of these selectors when requests are in progress or incomplete. So my suggestion would be to avoid the distinction between undefined and null here and always return null.

Reply:
I could have sworn that I'd been of the like opinion here that `undefined` is a good representation of "not yet known" and `null` of "known to be nothing". I'm not really sure why I'd reversed it here. I'd be agreeable to changing it though.

In general, I'd still agree it's a bit of an abuse of these values to try to signal the tri-state "unknown", "known and nothing", "known and something" via `undefined`, `null`, and some truthy value. I'm not sure the language gives us a great alternative here. I think there'd been cases where using some other value like `false` could be an issue (imagine something like `wasSaveSuccessful`, where `false` should mean "the save was not successful" and not "the save is still in progress and the status is not yet known"). Pretty sure there are real cases of this in the current code.